xrpl/models/base_model.py

```diff
@@ -284,20 +284,80 @@ def is_valid(self: BaseModel) -> bool:
         """
         return len(self._get_errors()) == 0
 
-    def _get_errors(self: BaseModel) -> Dict[str, str]:
+    def _check_type(
+        self: Self, attr: str, value: Any, expected_type: Type[Any]
+    ) -> Dict[str, str]:
+        # returns error message if type is bad, None if type is good
+        expected_type_origin = get_origin(expected_type)
+        if expected_type_origin is Union:
+            if any(
+                len(self._check_type(attr, value, expected_type_option)) == 0
+                for expected_type_option in get_args(expected_type)
+            ):
+                return {}
+            return {attr: f"{attr} is {type(value)}, expected {expected_type}"}
+
+        if expected_type is Any:
+            return {}
+
+        if expected_type_origin is list:
+            # expected a List, received a List
+            if not isinstance(value, list):
+                return {"attr": f"{attr} is {type(value)}, expected {expected_type}"}
+            result = {}
+            for i in range(len(value)):
+                result.update(
+                    self._check_type(
+                        f"{attr}[{i}]", value[i], get_args(expected_type)[0]
+                    )
+                )
+            return result
+
+        if expected_type_origin is dict:
+            return (
+                {}
+                if isinstance(value, dict)
+                else {attr: f"{attr} is {type(value)}, expected {expected_type}"}
+            )
+
+        if isinstance(expected_type, type) and issubclass(expected_type, Enum):
+            return (
+                {}
+                if value in list(expected_type)
+                else {
+                    attr: f"{attr} is {value}, expected member of {expected_type} enum"
+                }
+            )
+
+        if expected_type_origin is Literal:
+            arg = get_args(expected_type)
+            return {} if value in arg else {attr: f"{attr} is {value}, expected {arg}"}
+
+        if issubclass(expected_type, BaseModel) and isinstance(value, dict):
+            return {}
+
+        if not isinstance(value, expected_type):
```

Comment:
What about `Optional` type?  `return {} # last line of the function` will be hit for the `Optional` type. I think we'd want to validate it separately right?

Reply:
`Optional` is `Union[type, None]` under the hood.